// @framer/ruucm-h.facebook-video/code/Facebook_Video.tsx
import * as React from 'react'
import { PropertyControls, ControlType } from 'framer'
import styled, { css } from 'styled-components'
import { Helmet } from 'react-helmet'

const Wrap = styled.div`
  width: 100%;
  height: 100%;
  ${props =>
    props.backgroundColor &&
    css`
      background: ${props.backgroundColor};
    `}
`
const VideoWrap = styled.div`
  position: absolute !important;
  top: 50%;
  transform: translateY(-50%);
`

// Define type of property
interface Props {
  url: string
  width: number
  backgroundColor: string
}

export class Facebook_Video extends React.Component<Props> {
  // Set default properties
  static defaultProps = {
    url: 'https://www.facebook.com/facebook/videos/10153231379946729/"',
    backgroundColor: 'rgba(136, 85, 255, 0.1)',
  }

  // Items shown in property panel
  static propertyControls: PropertyControls = {
    url: { type: ControlType.String, title: 'Video URL' },
    backgroundColor: { type: ControlType.Color, title: 'Background Color' },
  }

  render() {
    return (
      <Wrap backgroundColor={this.props.backgroundColor}>
        <Helmet>
          {/* <!-- Load Facebook SDK for JavaScript --> */}
          <script>{`
      ;(function(d, s, id) {
        var js,
          fjs = d.getElementsByTagName(s)[0]
        if (d.getElementById(id)) return
        js = d.createElement(s)
        js.id = id
        js.src =
          'https://connect.facebook.net/ko_KR/sdk.js#xfbml=1&version=v3.2&appId=855675654445896&autoLogAppEvents=1'
        fjs.parentNode.insertBefore(js, fjs)
      })(document, 'script', 'facebook-jssdk')
    `}</script>
        </Helmet>

        <VideoWrap
          className="fb-video"
          data-href={this.props.url}
          data-width={this.props.width}
          data-show-text="false"
          data-autoplay="true"
        />
      </Wrap>
    )
  }
}
