// @framer/tryjohnny.wistia-video-player/code/Wistia.tsx

import * as React from "react";
import { Animatable, PropertyControls, ControlType } from "framer"

interface Props {
  source: string,
  playerColor: string,
  width: number,
  height: number,
  playerControls: boolean,
}

export class Wistia extends React.Component {
  static defaultProps = {
    source: "1ulz5u362b",
    playerColor: '54bbff',
    width: 1280,
    height: 720,
    playerControls: true,
  };

  static propertyControls: PropertyControls<Props> = {
    source: {
      type: ControlType.String,
      title: 'Wistia ID'
    },
    playerColor: {
      type: ControlType.String,
      title: 'Hex Code'
    },
    playerControls: {
      type: ControlType.Boolean,
      title: 'Controls?',
      enabled: "Yes",
			disabled: "No",
    }
  }

  render() {
    const width = Animatable.getNumber(this.props.width)
    const height = Animatable.getNumber(this.props.height)

    var url = `http://fast.wistia.com/embed/iframe/${this.props.source}?version=v1&controlsVisibleOnLoad=${this.props.playerControls}&playerColor=${this.props.playerColor}`;

    return (
        <iframe ref="iframe" width={width} height={height} frameBorder={0} src={url}/>
    )
  }
}
